### DISCUSSION TITLE
Add custom configuration to APIRoute

### DISCUSSION THREAD

**Discussion by shashankrnr32** [fastapi/fastapi#4140] [Questions] [answered]
First of all, Thank you very much for this amazing project. I have recently started learning and using FastAPI. So, If this question is answered, please feel free to mark it as duplicate and link it to the older question.

#### Question

Currently all the REST decorators (router.get, router.post, ...) allow the following params given in this [link](https://github.com/tiangolo/fastapi/blob/58ab733f19846b4875c5b79bfb1f4d1cb7f4823f/fastapi/routing.py#L479) to be configured. Is there a way to add a custom configuration (say a dictionary) to the route definition which can be accessed in a middleware to make certain decision about a particular route?

The usecase I am explaining here is for a middleware, but this can be used in different places.

#### My Usecase.

The prometheus middleware given [here](https://github.com/perdy/starlette-prometheus/blob/master/starlette_prometheus/middleware.py) loops through all the routes configured for an app and gets the full match. I want to add certain configuration to some of my APIs in my app which can be used in the middleware not to monitor an API.

For example;

```python
@router.get("/health", config=dict(monitor=False))
```

With the above `config` (and maybe other key-value pairs), my middleware can take decision not to monitor that particular API. Something in [this](https://github.com/tiangolo/fastapi/compare/master...shashankrnr32:route_config?expand=1) manner

#### Some workarounds

1. The method does have an `openapi_extra` parameter that can be used (as a workaround), but this will get to the schema of the API. And this is definitely not the right place to add such configuration

2. I am aware that I can extend these classes to add some more params in these methods, but want to get opinion here if this is something others want too.

Thank you!

**Comment by simondale00** [comment, marked as answer]
Just throwing in my two cents...

I wonder if this specific use-case is better addressed in the `starlette-prometheus` module. FastAPI supports passing configuration to the middleware when it's installed. It could work as follows:

```
app = FastAPI()
app.add_middleware(PrometheusMiddleware, exclude_paths=[
    "/health",
    "/some/other/path"
])
```

This pattern would allow `PrometheusMiddleware` to be explicitly configured as desired rather than having to rely on unstructured metadata attached to each route.

Best,
Simon

**Reply by shashankrnr32** [reply]
Reasoning behind selecting this as an answer: https://github.com/tiangolo/fastapi/discussions/4140#discussioncomment-1662919

**Comment by simondale00** [comment]
Quick update....

I wonder if the [sub application pattern](https://fastapi.tiangolo.com/advanced/sub-applications/?h=sub) would also help solve your problem. The health endpoint could be registered in a sub application with it's own middleware (or in this case, lack thereof).

Found the suggestion in a related question [here](https://github.com/tiangolo/fastapi/discussions/4027).

Best,
Simon

**Reply by shashankrnr32** [reply]
Thank you very much for the quick reply!

_Replying in this thread for both the answers_

I am currently using the pattern of `exclude_paths` of passing the excluded paths for this particular usecase. Currently it does serve the purpose. My thought was it would be easier to mark the API with a particular `label` (I could use `tags` for this particular case).

The sub application pattern sounds really interesting (and promising from the docs). I will definitely try this out for this particular usecase and also some other usecase that needed a similar approach. Once I have tried that out(in a week), if I see this as the better approach to achieve this, I will mark this as the answer. Until then, I will keep this open (to see if there are any other opinions).

> But I still do think its good to have some kind of configuration parameter that can be added to the `APIRoute` class. Maybe just my personal opinion 😄 

Thank you again!

**Reply by shashankrnr32** [reply]
Updating after 2 weeks

I did test out the `mount` option as well. It does work very well upto the expectations mentioned in my usecase. However I did not go with this approach as it requires me to logically separate out APIs that possibly belong in the same "Resource" module. This would not be a good design for me. So, I am currently keeping the same approach as to pass a tuple of predefined route templates to my middleware and ignoring if it matches.

This however might not scale up very well for different usecases if the configuration is more than a mere `boolean` (which I will worry then!!!).

So, I am marking the 1st answer as a recommended answer, although both of the above answers does solve the question.

Thank you!
